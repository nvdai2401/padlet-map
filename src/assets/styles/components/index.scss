@import './top-bar';
@import './folk-me';
@import './spinner';
@import './post-content';
@import './post-popup';
@import './mini-post';
@import './post-preview';
@import './post-expanded';
@import './map-control';
@import './map-view';
