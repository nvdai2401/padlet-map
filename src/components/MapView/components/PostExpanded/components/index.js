export { default as PostSlider } from './PostSlider';
