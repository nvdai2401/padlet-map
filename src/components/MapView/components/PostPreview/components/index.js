export { default as MiniPost } from './MiniPost';
