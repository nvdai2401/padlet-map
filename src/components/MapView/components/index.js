export { default as Post } from './Post';
export { default as PostPreview } from './PostPreview';
export { default as PostExpanded } from './PostExpanded';
