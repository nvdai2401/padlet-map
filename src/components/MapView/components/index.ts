export { default as PostContent } from './PostContent';
export { default as PostPopup } from './PostPopup';
export { default as PostPreview } from './PostPreview';
export { default as PostExpanded } from './PostExpanded';
export { default as MapControl } from './MapControl';
