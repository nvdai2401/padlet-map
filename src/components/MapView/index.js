export { default } from './MapView.vue';
