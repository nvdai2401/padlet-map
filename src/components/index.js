export { default as MapView } from './MapView';
export { default as TopBar } from './TopBar';
