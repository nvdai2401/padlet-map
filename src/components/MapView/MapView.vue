<template>
  <div class="map-view">
    <gmap-map
      ref="gmap"
      class="gmap-container"
      :zoom="initialZoom"
      :center="center"
      :options="mapOptions"
      @click="handleOnClickMap"
    >
      <gmap-marker
        ref="markers"
        v-for="marker in Object.values(markers)"
        :key="marker.id"
        :icon="{
          url: require(`src/assets/images/markers/${marker.color}_marker.webp`),
        }"
        :clickable="true"
        :position="{
          lat: marker.location_point.latitude,
          lng: marker.location_point.longitude,
        }"
        @click="handleOnClick(marker.id)"
        @mouseover="handleOnMouseOver(marker.id)"
        @mouseout="handleOnMouseOut(marker.id)"
      >
        <gmap-info-window
          :opened="marker.is_post_visible"
          @closeclick="is_post_visible = false"
        >
          <post
            v-if="marker.is_post_visible"
            :post-info="marker"
            :onExpandPost="handleOnExpandPost"
          />
        </gmap-info-window>
      </gmap-marker>
    </gmap-map>

    <div :class="['map-control', postPreviewVisible ? 'preview-shown' : '']">
      <button
        class="button map-center-control"
        @click="postPreviewVisible = !postPreviewVisible"
      >
        <font-awesome-icon :icon="['fas', 'bars']" />
      </button>

      <div class="group-button">
        <button class="button" @click="fitBounds">
          <font-awesome-icon :icon="['fas', 'expand']" />
        </button>
        <button class="button" @click="zoomIn">
          <font-awesome-icon :icon="['fas', 'plus']" />
        </button>
        <button class="button" @click="zoomOut">
          <font-awesome-icon :icon="['fas', 'minus']" />
        </button>
      </div>
    </div>
    <transition-group tag="div" name="slide">
      <post-preview
        v-show="postPreviewVisible && Object.keys(markers).length > 0"
        key="post-preview"
        :posts="markers"
        :onClick="handleOnClick"
        :onMouseOver="handleOnMouseOver"
        :onMouseOut="handleOnMouseOut"
      />
    </transition-group>

    <post-expanded
      v-if="postExpandedVisible"
      :open="postExpandedVisible"
      :postId="activePost"
      :posts="Object.keys(markers)"
      :onClose="handleOnClosePostExpanded"
    />
  </div>
</template>

<script>
import { Post, PostPreview, PostExpanded } from "./components";
import {
  defaultLatLngBounds,
  viewportLatLngBounds,
  options,
  DEFAULT_CENTER,
  MIN_ZOOM,
  MAX_ZOOM,
} from "./mapProperty";
import { getMarkers } from "src/api";

export default {
  name: "map-view",
  components: {
    post: Post,
    "post-preview": PostPreview,
    "post-expanded": PostExpanded,
  },
  data() {
    return {
      initialZoom: MIN_ZOOM,
      center: DEFAULT_CENTER,
      infoWindowVisible: false,
      postPreviewVisible: false,
      postExpandedVisible: false,
      markers: {},
      mapOptions: null,
      activePost: null,
      activeMarkerColor: "",
      focusedPost: "",
    };
  },
  watch: {
    activePost(newPostId, oldPostId) {
      const restriction = {
        ...this.mapOptions.restriction,
        latLngBounds: viewportLatLngBounds,
      };
      // Hide old post
      if (oldPostId) {
        this.markers[oldPostId].is_post_visible = false;
      }
      // Hide current post and update map bounds to fit the viewport when click its marker
      if (!newPostId) {
        this.mapOptions = { ...this.mapOptions, restriction };
        return;
      }
      // Show post and change map bounds to default when click a marker
      this.markers[newPostId].is_post_visible = true;
      restriction.latLngBounds = defaultLatLngBounds;
      this.mapOptions = { ...this.mapOptions, restriction };
    },
  },
  created() {
    this.mapOptions = options;
  },
  async mounted() {
    await this.fetchMakers();
    this.fitBounds();
  },
  methods: {
    async fetchMakers() {
      const data = await getMarkers();
      const markers = {};

      for (let item of data) {
        markers[item.id] = { ...item.attributes, is_post_visible: false };
      }
      this.markers = markers;
    },
    handleOnClick(postId) {
      if (this.activePost === postId) {
        this.activePost = null;
        return;
      }
      this.activePost = postId;
    },
    handleOnMouseOver(postId) {
      if (postId !== this.focusedPost) {
        const PLACEHOLDER_PREFIX = "ph_";
        this.focusedPost = postId;
        this.activeMarkerColor = this.markers[postId].color;
        this.markers[postId].color =
          PLACEHOLDER_PREFIX + this.markers[postId].color;
      }
    },
    handleOnMouseOut(postId) {
      this.markers[postId].color = this.activeMarkerColor;
      this.activeMarkerColor = "";
      this.focusedPost = "";
    },
    handleOnClickMap() {
      this.activePost = null;
    },
    fitBounds() {
      this.$refs.gmap.$mapPromise.then((map) => {
        var bounds = new google.maps.LatLngBounds();
        Object.values(this.markers)
          .reverse()
          .forEach((marker) => {
            bounds.extend(
              new google.maps.LatLng(
                marker.location_point.latitude,
                marker.location_point.longitude
              )
            );
          });
        map.fitBounds(bounds);
      });
    },
    handleOnExpandPost() {
      this.postExpandedVisible = true;
    },
    handleOnClosePostExpanded() {
      this.postExpandedVisible = false;
      this.activePost = null;
    },
    zoomIn() {
      const currentZoom = this.$refs.gmap.$mapObject.zoom;

      if (currentZoom < MAX_ZOOM) {
        this.mapOptions = { ...this.mapOptions, zoom: currentZoom + 1 };
      }
    },
    zoomOut() {
      const currentZoom = this.$refs.gmap.$mapObject.zoom;

      if (currentZoom > MIN_ZOOM) {
        this.mapOptions = { ...this.mapOptions, zoom: currentZoom - 1 };
      }
    },
  },
};
</script>

<style lang="scss" scoped>
.map-view {
  height: 90vh;
  position: relative;

  .gmap-container {
    width: 100vw;
    height: 100%;
  }

  .preview-shown {
    transform: translateY(-74px);
  }

  .map-control {
    position: absolute;
    z-index: 3000;
    bottom: 48px;
    left: 12px;
    display: flex;
    flex-direction: column;

    transition: transform 0.25s cubic-bezier(0.4, 0, 0.2, 1);

    svg {
      height: 14px;
    }

    .button {
      width: 30px;
      height: 30px;
      display: flex;
      align-items: center;
      justify-content: center;
      color: #666666;
      background-color: #ffffff;
      transition: all 0.15s ease-out;
      cursor: pointer;
      border: 0;
      outline: unset;

      &:hover {
        svg {
          color: #000000;
        }
      }
    }

    .map-center-control {
      border-radius: 6px;
    }

    .group-button {
      margin-top: 6px;
      border-radius: 6px;
      overflow: hidden;
      box-shadow: 0 1px 1px rgba(0, 0, 0, 0.12);
      transition: all 0.1s cubic-bezier(0.4, 0, 0.2, 1);
    }
  }

  .slide-leave-active,
  .slide-enter-active {
    transition: 0.3s;
  }
  .slide-enter {
    transform: translate(0, 100%);
  }
  .slide-leave-to {
    transform: translate(0, 100%);
  }

  @media screen and (min-width: 768px) {
    .preview-shown {
      transform: unset;
    }

    .map-control {
      flex-direction: row;
      bottom: 26px;
      right: 12px;
      left: unset;

      svg {
        height: 14px;
      }

      .group-button {
        display: flex;
        margin-top: 0;
        margin-left: 12px;
      }
    }

    .slide-enter {
      transform: translate(100%, 0);
    }
    .slide-leave-to {
      transform: translate(100%, 0);
    }
  }
}
</style>
