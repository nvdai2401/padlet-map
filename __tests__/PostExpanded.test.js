/* eslint-disable no-undef */
import { posts } from './shares/mockData';
import { mountWithProps, dispatchKeyDownEvent } from './shares/utils';
import { PostExpanded } from '@/components/MapView/components';

const propsData = {
  postId: posts[0].id,
  posts: posts.map((post) => post.id),
  onClose: jest.fn(),
};

describe('<PostExpanded />', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mountWithProps(PostExpanded, propsData);
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it('is a Vue instance', () => {
    expect(wrapper.isVueInstance).toBeTruthy();
  });

  it('should match snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it(`should display loading when fetching data`, async () => {
    wrapper.setData({ postInfo: posts[0], loading: true });
    expect(wrapper.find('.spinner').exists()).toBeTruthy();
  });

  it(`should contain 4 navigation buttons and 1 close button`, async () => {
    wrapper.setData({ postInfo: posts[0], loading: false });

    wrapper.vm.$nextTick(() => {
      expect(wrapper.findAll('button').length).toBe(5);
      expect(wrapper.findAll('svg').length).toBe(5);
    });
  });

  it(`should contain post information`, async () => {
    wrapper.setData({ postInfo: posts[0], loading: false });

    wrapper.vm.$nextTick(() => {
      expect(wrapper.find('.post-header').text()).toBe(posts[0].headline);
      expect(wrapper.find('.post-body').exists()).toBeTruthy();
      expect(wrapper.find('img').exists()).toBeTruthy();
    });
  });

  it(`should set currentIndex 0 when click Step backward button`, () => {
    wrapper.setData({ postInfo: posts[0], loading: false, currentIndex: 1 });
    wrapper
      .findAll('button')
      .at(0)
      .trigger('click');
    wrapper.vm.$nextTick(() => {
      expect(wrapper.vm.currentIndex).toBe(0);
    });
  });

  it(`should set currentIndex ${posts.length -
    1} when click Step backward button`, () => {
    wrapper.setData({ postInfo: posts[0], loading: false, currentIndex: 1 });
    wrapper
      .findAll('button')
      .at(3)
      .trigger('click');
    wrapper.vm.$nextTick(() => {
      expect(wrapper.vm.currentIndex).toBe(posts.length - 1);
    });
  });

  it(`should decrease currentIndex 1 when click Prev button`, () => {
    wrapper.setData({ postInfo: posts[0], loading: false, currentIndex: 1 });
    wrapper
      .findAll('button')
      .at(1)
      .trigger('click');
    wrapper.vm.$nextTick(() => {
      expect(wrapper.vm.currentIndex).toBe(0);
    });
  });

  it(`should decrease currentIndex 1 when press Arrow left key`, () => {
    const ARROW_LEFT_KEYCODE = 37;
    wrapper.setData({ postInfo: posts[0], loading: false, currentIndex: 1 });
    dispatchKeyDownEvent('keydown', ARROW_LEFT_KEYCODE);
    wrapper.vm.$nextTick(() => {
      expect(wrapper.vm.currentIndex).toBe(0);
    });
  });

  it(`should increase currentIndex 1 when click Next button`, () => {
    wrapper.setData({ postInfo: posts[0], loading: false, currentIndex: 1 });
    wrapper
      .findAll('button')
      .at(2)
      .trigger('click');
    wrapper.vm.$nextTick(() => {
      expect(wrapper.vm.currentIndex).toBe(2);
    });
  });

  it(`should increase currentIndex 1 when press Arrow right key`, () => {
    const ARROW_RIGHT_KEYCODE = 39;
    wrapper.setData({ postInfo: posts[0], loading: false, currentIndex: 1 });
    dispatchKeyDownEvent('keydown', ARROW_RIGHT_KEYCODE);
    wrapper.vm.$nextTick(() => {
      expect(wrapper.vm.currentIndex).toBe(2);
    });
  });

  it(`should call onClose when click Close button`, () => {
    wrapper
      .findAll('button')
      .at(4)
      .trigger('click');
    wrapper.vm.$nextTick(() => {
      expect(propsData.onClose).toHaveBeenCalled();
    });
  });
});
