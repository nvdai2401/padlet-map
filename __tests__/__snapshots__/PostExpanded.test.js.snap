// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PostExpanded /> should match snapshot 1`] = `
<div
  class="flex-column align-center post-expanded"
>
  <div
    class="flex-row align-center justify-space-between toolbar"
  >
    <div
      class="post-index"
    >
      0/13
    </div>
     
    <div
      class="nav-buttons"
    >
      <button>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-step-backward fa-w-14"
          data-icon="step-backward"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            class=""
            d="M64 468V44c0-6.6 5.4-12 12-12h48c6.6 0 12 5.4 12 12v176.4l195.5-181C352.1 22.3 384 36.6 384 64v384c0 27.4-31.9 41.7-52.5 24.6L136 292.7V468c0 6.6-5.4 12-12 12H76c-6.6 0-12-5.4-12-12z"
            fill="currentColor"
          />
        </svg>
      </button>
       
      <button>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-chevron-left fa-w-10"
          data-icon="chevron-left"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            class=""
            d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
            fill="currentColor"
          />
        </svg>
      </button>
       
      <button>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-chevron-right fa-w-10"
          data-icon="chevron-right"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 320 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            class=""
            d="M285.476 272.971L91.132 467.314c-9.373 9.373-24.569 9.373-33.941 0l-22.667-22.667c-9.357-9.357-9.375-24.522-.04-33.901L188.505 256 34.484 101.255c-9.335-9.379-9.317-24.544.04-33.901l22.667-22.667c9.373-9.373 24.569-9.373 33.941 0L285.475 239.03c9.373 9.372 9.373 24.568.001 33.941z"
            fill="currentColor"
          />
        </svg>
      </button>
       
      <button>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-step-forward fa-w-14"
          data-icon="step-forward"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 448 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            class=""
            d="M384 44v424c0 6.6-5.4 12-12 12h-48c-6.6 0-12-5.4-12-12V291.6l-195.5 181C95.9 489.7 64 475.4 64 448V64c0-27.4 31.9-41.7 52.5-24.6L312 219.3V44c0-6.6 5.4-12 12-12h48c6.6 0 12 5.4 12 12z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
     
    <div>
      <button>
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-times fa-w-11"
          data-icon="times"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 352 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            class=""
            d="M242.72 256l100.07-100.07c12.28-12.28 12.28-32.19 0-44.48l-22.24-22.24c-12.28-12.28-32.19-12.28-44.48 0L176 189.28 75.93 89.21c-12.28-12.28-32.19-12.28-44.48 0L9.21 111.45c-12.28 12.28-12.28 32.19 0 44.48L109.28 256 9.21 356.07c-12.28 12.28-12.28 32.19 0 44.48l22.24 22.24c12.28 12.28 32.2 12.28 44.48 0L176 322.72l100.07 100.07c12.28 12.28 32.2 12.28 44.48 0l22.24-22.24c12.28-12.28 12.28-32.19 0-44.48L242.72 256z"
            fill="currentColor"
          />
        </svg>
      </button>
    </div>
  </div>
   
  <div
    class="flex-column content"
  >
    <div
      class="flex-row align-center justify-center spinner"
    >
      <div
        class="loading"
      />
    </div>
  </div>
</div>
`;
